pkg/splunk/enterprise/monitoringconsole.go

```diff
@@ -16,17 +16,157 @@ package enterprise
+	if err != nil {
+		return result, err
+	}
+
+	// check if deletion has been requested
+	if cr.ObjectMeta.DeletionTimestamp != nil {
+		DeleteOwnerReferencesForResources(client, cr, &cr.Spec.SmartStore)
+		terminating, err := splctrl.CheckForDeletion(cr, client)
+		if terminating && err != nil { // don't bother if no error, since it will just be removed immmediately after
+			cr.Status.Phase = splcommon.PhaseTerminating
+		} else {
+			result.Requeue = false
+		}
+		return result, err
+	}
+
+	// create or update a headless service
+	err = splctrl.ApplyService(client, getSplunkService(cr, &cr.Spec.CommonSplunkSpec, SplunkMonitoringConsole, true))
+	if err != nil {
+		return result, err
+	}
+
+	// create or update a regular service
+	err = splctrl.ApplyService(client, getSplunkService(cr, &cr.Spec.CommonSplunkSpec, SplunkMonitoringConsole, false))
+	if err != nil {
+		return result, err
+	}
+
+	// create or update statefulset
+	statefulSet, err := getMonitoringConsoleStatefulSet(client, cr)
+	if err != nil {
+		return result, err
+	}
+
+	mgr := splctrl.DefaultStatefulSetPodManager{}
+	phase, err := mgr.Update(client, statefulSet, 1)
+	if err != nil {
+		return result, err
+	}
+	cr.Status.Phase = phase
+
+	// no need to requeue if everything is ready
+	if cr.Status.Phase == splcommon.PhaseReady {
+		result.Requeue = false
+	}
+	return result, nil
+}
+
+// getMonitoringConsoleStatefulSet returns a Kubernetes StatefulSet object for Splunk Enterprise monitoring console instances.
+func getMonitoringConsoleStatefulSet(client splcommon.ControllerClient, cr *enterprisev1.MonitoringConsole) (*appsv1.StatefulSet, error) {
+	// get generic statefulset for Splunk Enterprise objects
+	var monitoringConsoleConfigMap *corev1.ConfigMap
+	configMap := GetSplunkMonitoringconsoleConfigMapName(cr.GetName(), SplunkMonitoringConsole)
+	ss, err := getSplunkStatefulSet(client, cr, &cr.Spec.CommonSplunkSpec, SplunkMonitoringConsole, 1, []corev1.EnvVar{})
+	if err != nil {
+		return nil, err
+	}
+	//use mc configmap as EnvFrom source
+	ss.Spec.Template.Spec.Containers[0].EnvFrom = []corev1.EnvFromSource{
+		{
+			ConfigMapRef: &corev1.ConfigMapEnvSource{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: configMap, //monitoring console env variables configMap
+				},
+			},
+		},
+	}
+
+	//update podTemplate annotation with configMap resource version
+	namespacedName := types.NamespacedName{Namespace: cr.GetNamespace(), Name: configMap}
+	monitoringConsoleConfigMap, err = splctrl.GetMCConfigMap(client, namespacedName)
+	if err != nil {
+		return nil, err
+	}
+
+	//set the owner reference for the corresponding configmap
+	//TODO: this will delete the cm once the mc pod is deleted, once mcref is defined in other CRs verify
+	//TODO: if same mcref is present in some CRs then cm if created
+	err = splctrl.SetConfigMapOwnerRef(client, cr, namespacedName)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Can we move this to the configmap.go, block starting at line #84, so that, we don't call it only once, and do not duplicate across the reconciles.